src/hotspot/share/cds/filemap.cpp

```diff
@@ -1117,10 +1117,15 @@ bool FileMapInfo::validate_shared_path_table() {
     assert(shared_path(0)->is_modules_image(), "first shared_path must be the modules image");
   } else {
     if (!validate_boot_class_paths() || !validate_app_class_paths(shared_app_paths_len)) {
-      const char* mismatch_msg = "shared class paths mismatch (hint: enable -Xlog:class+path=info to diagnose the failure)";
-      fail_continue("%s", mismatch_msg);
+      const char* mismatch_msg = "shared class paths mismatch";
+      const char* hint_msg = "(hint: enable -Xlog:class+path=info to diagnose the failure)";
+      if (log_is_enabled(Info, class, path)) {
+        fail_continue("%s", mismatch_msg);
+      } else {
+        fail_continue("%s %s", mismatch_msg, hint_msg);
+      }
       if (!log_is_enabled(Info, cds) && !log_is_enabled(Info, class, path)) {
-        log_warning(cds)("%s", mismatch_msg);
+        log_warning(cds)("%s %s", mismatch_msg, hint_msg);
```

Comment:
Usually, the message is printed inside `fail_continue()`, which goes to `log_info()`. However, in this particular case, we feel that the message is important enough that it should be printed in `log_warning()`. The current implementation is kind of awkward and makes this PR complicated. 

I think we can take this chance to clean things up. We can add an argument to `fail_continue()` to select the logging level. Something like this:

```
void FileMapInfo::fail_continue(const char *msg, ...) {
  va_list ap;
  va_start(ap, msg);
  fail_continue_impl(LogLevel::Info, msg, ap);
}

void FileMapInfo::fail_continue(LogLevelType level, const char *msg, ...) {
  va_list ap;
  va_start(ap, msg);
  fail_continue_impl(level, msg, ap);
}

void FileMapInfo::fail_continue_impl(LogLevelType level, const char *msg, va_list ap) {
...

-     if (log_is_enabled(Info, cds)) {
-       LogStream ls(Log(cds)::info());
-       ls.print("UseSharedSpaces: "); 
-       ls.vprint_cr(msg, ap);
-     }
+     LogMessage(cds) lm;
+     lm.vwrite(level, msg, ap);
...
}
```

And thus this function can be simplified as:

```
const char* mismatch_msg = "shared class paths mismatch";
const char* hint_msg = log_is_enabled(Info, class, path) ?
    "" : " (hint: enable -Xlog:class+path=info to diagnose the failure)";
fail_continue(LogLevel::warning, "%s%s", mismatch_msg, hint_msg);
```

The "UseSharedSpaces:" is also redundant since the log already contains `[cds]`. It's hard to pass that to `LogMessage` so we may as well just remove it.

Reply:
I've removed the "UseSharedSpaces:" log message and adjusted a few tests.

Reply:
I've simplified the `FileMapInfo::fail_continue_impl`by removing the va_* calls.